File: lulu/first_position_of_target.py
class Solution:
    # @param nums: The integer array
    # @param target: Target number to find
    # @return the first position of target in nums, position start from 0 
    def binarySearch(self, nums, target):
        # write your code here
        return self.search(nums, 0, len(nums)-1, target)
        
    def search(self, nums, start, end, target):
        
        if start == end:
            if nums[start] == target:
                return start
            else:
                return -1
        i = (start + end) / 2
        if nums[i] >= target:
            return self.search(nums, start, i, target)
        elif nums[i] < target:
            return self.search(nums, i+1, end, target)

File: bsdelf/first-position-of-target.cc
/**
 * @author      Yanhui Shen
 * @version     1.0
 * @since       2016-02-11 
 */

class Solution {
public:
    /**
     * @param nums: The integer array.
     * @param target: Target number to find.
     * @return: The first position of target. Position starts from 0. 
     */
    int binarySearch(vector<int> &array, int target) {
        size_t first = 0;
        int64_t count = array.size();
        while (count > 0) {
            size_t pos = first; 
            size_t step = count / 2; 
            pos += step;
            if (array[pos] < target) {
                first = ++pos;
                count -= step + 1; 
            } else {
                count = step;
            }
        }
        return array[first] == target ? first : -1; 
    }
};
